src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java

```diff
@@ -776,60 +773,33 @@ public String getAuthenticationGatewayUrl() {
-                                grantedAuthorities.add(SecurityRealm.AUTHENTICATED_AUTHORITY2);
-                                if (isNotBlank(escapeHatchGroup)) {
-                                    grantedAuthorities.add(new SimpleGrantedAuthority(escapeHatchGroup));
-                                }
-                                UsernamePasswordAuthenticationToken token = new UsernamePasswordAuthenticationToken(
-                                        escapeHatchUsername, "", grantedAuthorities);
-                                SecurityContextHolder.getContext().setAuthentication(token);
-                                OicUserDetails userDetails =
-                                        new OicUserDetails(escapeHatchUsername, grantedAuthorities);
-                                SecurityListener.fireAuthenticated2(userDetails);
-                                return token;
-                            } else {
-                                throw new BadCredentialsException("Wrong username and password: " + authentication);
-                            }
+        return new SecurityComponents(new AuthenticationManager() {
+            public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+                if (authentication instanceof AnonymousAuthenticationToken) return authentication;
+
+                if (authentication instanceof UsernamePasswordAuthenticationToken && escapeHatchEnabled) {
+                    randomWait(); // to slowdown brute forcing
+                    if (checkEscapeHatch(
+                            authentication.getPrincipal().toString(),
+                            authentication.getCredentials().toString())) {
+                        List<GrantedAuthority> grantedAuthorities = new ArrayList<>();
+                        grantedAuthorities.add(SecurityRealm.AUTHENTICATED_AUTHORITY2);
+                        if (isNotBlank(escapeHatchGroup)) {
+                            grantedAuthorities.add(new SimpleGrantedAuthority(escapeHatchGroup));
                         }
-                        throw new BadCredentialsException("Unexpected authentication type: " + authentication);
+                        UsernamePasswordAuthenticationToken token =
+                                new UsernamePasswordAuthenticationToken(escapeHatchUsername, "", grantedAuthorities);
+                        SecurityContextHolder.getContext().setAuthentication(token);
+                        OicUserDetails userDetails = new OicUserDetails(escapeHatchUsername, grantedAuthorities);
+                        SecurityListener.fireAuthenticated2(userDetails);
+                        return token;
+                    } else {
+                        throw new BadCredentialsException("Wrong username and password: " + authentication);
                     }
-                },
-                new UserDetailsService() {
```

Comment:
Switching to `SecurityComponents(AuthenticationManager)` constructor. `UserDetailsServiceProxy` throws `UserMayOrMayNotExistException2`; `ImpersonatingUserDetailsService2` handles that by checking `LastGrantedAuthoritiesProperty`.